### ISSUE TITLE
(Python) Attribute error when nose tries to print error

### ISSUE THREAD

**Issue by JannKleen** [rethinkdb/rethinkdb#4764]
When running tests which raise an error from rethinkdb, nose is unable to print the error and gives the following stacktrace instead:

```
Traceback (most recent call last):
File "/Users/xxx/.virtualenvs/yyy/bin/nosetests", line 11, in <module>
sys.exit(run_exit())
File "/Users/xxx/.virtualenvs/yyy/lib/python2.7/site-packages/nose/core.py", line 121, in __init__
**extra_args)
File "/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/main.py", line 95, in __init__
self.runTests()
File "/Users/xxx/.virtualenvs/yyy/lib/python2.7/site-packages/nose/core.py", line 207, in runTests
result = self.testRunner.run(self.test)
File "/Users/xxx/.virtualenvs/yyy/lib/python2.7/site-packages/nose/core.py", line 66, in run
result.printErrors()
File "/Users/xxx/.virtualenvs/yyy/lib/python2.7/site-packages/nose/result.py", line 103, in printErrors
_TextTestResult.printErrors(self)
File "/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/runner.py", line 108, in printErrors
self.printErrorList('ERROR', self.errors)
File "/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/runner.py", line 114, in printErrorList
self.stream.writeln("%s: %s" % (flavour,self.getDescription(test)))
File "/Users/xxx/.virtualenvs/yyy/lib/python2.7/site-packages/nose/result.py", line 82, in getDescription
return test.shortDescription() or str(test)
File "/Users/xxx/.virtualenvs/yyy/lib/python2.7/site-packages/nose/case.py", line 51, in __str__
return str(self.test)
File "/Users/xxx/.virtualenvs/yyy/lib/python2.7/site-packages/nose/failure.py", line 31, in __str__
getattr(self.exc_class, '__name__', self.exc_class), self.exc_val)
File "/Users/xxx/.virtualenvs/yyy/lib/python2.7/site-packages/rethinkdb/errors.py", line 72, in __str__
self.query_printer.print_carrots()))
File "/Users/xxx/.virtualenvs/yyy/lib/python2.7/site-packages/rethinkdb/errors.py", line 34, in convertForPrint
return inputString.encode(sys.stdout.encoding or 'utf-8',
                          AttributeError: StringIO instance has no attribute 'encoding'
```

**Comment by JannKleen** [comment]
Just realised that this might be a duplicate of #4114

**Comment by larkost** [comment]
From the look of this it is because you are trying to use [`nose`](https://nose.readthedocs.org/en/latest/) as a replacement to the built-in `unittest` and that is trying to be clever at one point and use a StringIO object where we don't support that.

Probably the fundamental problem here is that we don't use `unitest` as our basic testing harness. It is used in some Python tests, but that is more for my convenience than anything else (remember, we have a number of languages to test). Rather you should be using our two testing harnesses to run our tests:
- `./test/run` this currently runs our server unitests, and a number of "scenario" tests that do different things with the server
- `./test/rql_test/test-runner` runs our client driver tests (we call them "polyglot" and "connection" tests)

It is worth noting that I have mid-term plans to merge these two systems in #3442, at which time there will only be one test harness.

On the assumption that this is the actual problem, I am going to close this. Please comment if this is a bad assumption.

**Comment by larkost** [comment]
I just re-read this, and am re-opening. I was initially reading this as using our tests, but I think you are running your own tests that happen to use the RethinkDB python driver. Sorry about missing that.

**Comment by JannKleen** [comment]
yes, correct :+1: any idea how I can provide you any additional information or help?
I don't want to exclude the possibility that I'm making a stupid mistake here, but I can't really see where :smile:

**Comment by larkost** [comment]
The most helpful thing would be to distill the failing test to the smallest reproducible case, and the include that here, or if that is not practical [email it to me](mailto://kalr@rethinkdb.com). Unless you know you are using a StringIO object (my guess is that `nose` is doing that for you), then tell us how you are using that.

I can't promise that this will be high-priority, but with a straightforward test case in front of one of us it might well be an easy fix.

**Comment by JannKleen** [comment]
I get it's not a high priority issue, I might start working on a PR myself if this is indeed an issue with rethinkdb and not my fault.

https://gist.github.com/JannKleen/b7acad567ada75c88edb

These tests cause similar errors:
```
======================================================================
ERROR: tests.test_something
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/xxx/.virtualenvs/resmio/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/private/tmp/yyy/tests.py", line 5, in test_something
    r.db("test").table_create("authors").run()
  File "/Users/xxx/.virtualenvs/resmio/lib/python2.7/site-packages/rethinkdb/ast.py", line 118, in run
    return c._start(self, **global_optargs)
  File "/Users/xxx/.virtualenvs/resmio/lib/python2.7/site-packages/rethinkdb/net.py", line 575, in _start
    return self._instance.run_query(q, global_optargs.get('noreply', False))
  File "/Users/xxx/.virtualenvs/resmio/lib/python2.7/site-packages/rethinkdb/net.py", line 450, in run_query
    raise res.make_error(query)
ReqlOpFailedError: <unprintable ReqlOpFailedError object>

======================================================================
ERROR: tests.test_another_thing
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/xxx/.virtualenvs/resmio/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/private/tmp/yyy/tests.py", line 13, in test_another_thing
    print e
  File "/Users/xxx/.virtualenvs/resmio/lib/python2.7/site-packages/rethinkdb/errors.py", line 72, in __str__
    self.query_printer.print_carrots()))
  File "/Users/xxx/.virtualenvs/resmio/lib/python2.7/site-packages/rethinkdb/errors.py", line 34, in convertForPrint
    return inputString.encode(sys.stdout.encoding or 'utf-8',
AttributeError: StringIO instance has no attribute 'encoding'

----------------------------------------------------------------------
Ran 2 tests in 0.032s
```

**Comment by larkost** [comment]
It was about what I guessed: `nose` is being clever and using `StringIO` as a replacement for `sys.stdout` (I would have gone with a `NamedTemporaryFile`) and this does not have an `encoding` property. So the driver is tripping over that when trying to output. A fix for this is in db00729f.

@danielmewes: should I cherry-pick this to `v2.1.x`?

**Comment by danielmewes** [comment]
@larkost Cherry-picking into `v2.1.x` sounds like a good idea.

**Comment by danielmewes** [comment]
@larkost Do you think this is the same as https://github.com/rethinkdb/rethinkdb/issues/4114 btw?

**Comment by larkost** [comment]
@danielmewes yes, this was the same thing. I have closed that one.

**Comment by larkost** [comment]
This has been cherry-picked into `v2.1.x` with 5bf573dd

**Comment by JannKleen** [comment]
:trophy: awesome, thanks for the quick fix!

**Comment by danielmewes** [comment]
The fix was shipped with the 2.1.0.post1 Python driver.
